### ISSUE TITLE
Cannot assign `cls.some_dict: [Hashable, t.Self]` in `__init_subclass__`

### ISSUE THREAD

**Issue by aThorp96** [microsoft/pyright#9011]
**Describe the bug**
If you have a class variable which holds objects of type `Self`, then Pyright will not allow you to reassign the variable in a class method with a seemingly wrong error message: `Type parameter "_VT@dict" is invariant, but "Self@Parent" is not the same as "Self@Parent"`

**Code or Screenshots**
Repro in the pyright playground: [link](https://pyright-play.net/?pythonVersion=3.12&strict=true&code=JYWwDg9gTgLgBDAnmYA7A5nAhgZwQKHwGMAbXPABSygFNUYAufOFuHGAVwDMuGEA6AMJkcOAGrUA2gBNgRGJPZQANAIDKNElwC62wqzgABUuRA0YACwjTmraTS5wA%2Bk7TAYTnBwBGJ0S4AKUhwASiYDA2D%2Bdm5HAF44AG8AX0I-PABpYGkARgCqWnowuH5S4hFM7IAmfOo6GGLS-kIs3OjOHkkAIhwIMycAaxpELu04BNa8kPxWqvbY7t7%2BoZGxieqA6aA)
Including here for posterity as well:
```python
import typing as t

class Parent:
    stuff: t.ClassVar[dict[str, t.Self]]

    @classmethod
    def __init_subclass__(cls):
        cls.stuff = {}

class Kid1(Parent): ...
class Kid2(Parent): ...

Kid1.stuff["some key"] = Kid1()
Kid2.stuff["some key"] = Kid2()
```

**VS Code extension or command-line**
This appears to be the case in the Command Line Language Server as well as the pyright playground.

**Comment by erictraut** [comment]
Thanks for the bug report. There were two separate bugs here, both of which I've fixed. They will both be addressed in the next release of pyright.

Bug 1: An incorrect error message extension was being appended to the error regarding a missing `__set__` method. This doesn't apply in this case because the attribute being accessed is not a descriptor object.

Bug 2: When inferring the type of `{}`, pyright uses the declared type of the LHS (the expression `cls.suff`), but it was not properly binding the type of `stuff` to `cls` when evaluating the LHS type. This resulted in an incorrect type inference for `{}` which could then not be assigned back to `cls.stuff`.

It's worth noting that your use of `Self` in a class variable type definition here is not strictly type safe. It should probably be flagged as a type error by type checkers for the same reason that other class-scoped type variables are not allowed in class variable types.

```python
class Foo[T]:
    x: ClassVar[T]  # Type error
```

`Self` is effectively a class-scoped type variable, no different from `T` in the above example. 

It's unsafe to use class-scoped TypeVars in class variables because all instances of a class share the same class object regardless of their specialization. That means a `Foo[int]` and a `Foo[str]` share the same `Foo.x`, so any type you ascribe to `x` needs to be valid for all instances of `Foo`. For that same reason, `Self` is generally unsafe to use in a `ClassVar` definition. I think your usage here is safe but only because you're assigning a new value to this class variable for each subclass.

The typing spec doesn't currently prohibit the use of `Self` in this context, but it probably should. This is an area of ambiguity that is on my list of issues to clarify in the spec. When it is, you may find that this code no longer type checks without errors. I mention this in case you want to proactively eliminate the use of `Self` in this context.

**Comment by aThorp96** [comment]
Thank you so much for the quick reply and fix @erictraut! And thanks for the insight around `Self` usage as well.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.381.
